File: gearshift2/src/main/java/org/sugr/gearshift/viewmodel/databinding/PropertyChangedDebouncer.java
package org.sugr.gearshift.viewmodel.databinding;

import android.databinding.Observable;

import java.util.concurrent.TimeUnit;

import rx.Subscription;
import rx.functions.Action1;
import rx.subjects.PublishSubject;
import rx.subjects.Subject;

public class PropertyChangedDebouncer extends Observable.OnPropertyChangedCallback {
    private final Subject<Observable, Observable> subject = PublishSubject.create();

    @Override
    public void onPropertyChanged(Observable observable, int i) {
        subject.onNext(observable);
    }

    public Subscription subscribe(Action1<Observable> onNext) {
        return debouncer().subscribe(onNext);
    }

    public Subscription subscribe(Action1<Observable> onNext, Action1<Throwable> onError) {
        return debouncer().subscribe(onNext, onError);
    }

    private rx.Observable<Observable> debouncer() {
        return subject.debounce(300, TimeUnit.MILLISECONDS);
    }
}

File: gearshift2/src/main/kotlin/org/sugr/gearshift/viewmodel/databinding/PropertyChangedDebouncer.kt
package org.sugr.gearshift.viewmodel.databinding

import android.databinding.Observable
import rx.functions.Action1
import rx.internal.util.InternalObservableUtils
import rx.lang.kotlin.PublishSubject
import java.util.concurrent.TimeUnit

class PropertyChangedDebouncer(
        private val transformer: rx.Observable.Transformer<Observable, Observable> =
            rx.Observable.Transformer {o -> o}
) : Observable.OnPropertyChangedCallback() {
    private val subject = PublishSubject<Observable>()

    override fun onPropertyChanged(observable: Observable, i: Int) = subject.onNext(observable)

    fun subscribe(
            onNext: Action1<Observable>,
            onError: Action1<Throwable> = InternalObservableUtils.ERROR_NOT_IMPLEMENTED
    ) = debouncer().subscribe(onNext, onError)

    fun subscribeChain(
            onNext: Action1<Observable>,
            onError: Action1<Throwable> = InternalObservableUtils.ERROR_NOT_IMPLEMENTED
    ) : PropertyChangedDebouncer{
        debouncer().subscribe(onNext, onError)
        return this;
    }

    private fun debouncer() = subject.debounce(300, TimeUnit.MILLISECONDS).compose(transformer)
}
